// azure-devops-node-api/interfaces/CIXInterfaces.js
/*
 * ---------------------------------------------------------
 * Copyright(C) Microsoft Corporation. All rights reserved.
 * ---------------------------------------------------------
 *
 * ---------------------------------------------------------
 * Generated file, DO NOT EDIT
 * ---------------------------------------------------------
 */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeInfo = exports.BuildFrameworkDetectionType = void 0;
const TfsCoreInterfaces = require("../interfaces/CoreInterfaces");
var BuildFrameworkDetectionType;
(function (BuildFrameworkDetectionType) {
    BuildFrameworkDetectionType[BuildFrameworkDetectionType["Shallow"] = 0] = "Shallow";
    BuildFrameworkDetectionType[BuildFrameworkDetectionType["Full"] = 10] = "Full";
})(BuildFrameworkDetectionType = exports.BuildFrameworkDetectionType || (exports.BuildFrameworkDetectionType = {}));
exports.TypeInfo = {
    BuildFrameworkDetectionType: {
        enumValues: {
            "shallow": 0,
            "full": 10
        }
    },
    CreatePipelineConnectionInputs: {},
};
exports.TypeInfo.CreatePipelineConnectionInputs.fields = {
    project: {
        typeInfo: TfsCoreInterfaces.TypeInfo.TeamProject
    }
};
